asn1/Stream.asn

```diff
@@ -0,0 +1,153 @@
+Stream
+DEFINITIONS
+AUTOMATIC TAGS ::=
+BEGIN
+
+IMPORTS
+    UInt8,
+    UInt96,
+    UInt128,
+    VarUInt,
+    VarBytes
+    FROM GenericTypes
+
+    Address
+    FROM InterledgerTypes
+;
+
+StreamEncryptionEnvelope ::= SEQUENCE {
+    -- Random nonce
+    initializationVector UInt96,
+
+    -- Authentication tag (output of AES-GCM encryption)
+    authenticationTag UInt128,
+
+    -- Encrypted STREAM Packet
+    -- Note: This is NOT encoded as a variable length (length-prefixed)
+    -- octet string. The size is constrained by the length of the data
+    -- field in the InterledgerPrepare, InterledgerFulfill, or InterledgerReject
+    -- packet containing this Psk2EncryptionEnvelope
+    cipherText OCTET STRING (SIZE (0..32739))
+}
+
+StreamPacket ::= SEQUENCE {
+  -- Always 1 for now
+  version UInt8,
+
+  -- ILP Packet type this STREAM packet must be carried upon
+  ilpPacketType UInt8,
+
+  -- Packet ID
+  sequence VarUInt,
+
+  -- If attached to a Prepare packet, minimum amount that the receiver should accept
+  -- If attached to a Fulfill or Reject packet, amount that arrived at the receiver
+  prepareAmount VarUInt,
+
+  frames SEQUENCE OF Frame
+}
+
+FRAME ::= CLASS {
+  &typedId UInt8 UNIQUE,
+  &Type
+} WITH SYNTAX {&typeId &Type}
+
+FrameSet FRAME ::= {
+  -- Connection-related frames (hex 0x0?)
+  {1 ConnectionError} |
+  --{2 ApplicationError} |
+  --{3 ConnectionMaxMoney} |
+  --{4 ConnectionMoneyBlocked} |
+  --{5 ConnectionMaxData} |
+  --{6 ConnectionDataBlocked} |
+  --{7 ConnectionMaxStreamId} |
+  --{8 ConnectionStreamIdBlocked} |
+  {9 ConnectionNewAddress} |
+
+  -- MoneyStream-related frames (hex 0x1?)
+  {16 StreamMoney} |
+  {17 StreamMoneyEnd} |
+  {18 StreamMoneyMax} |
+  --{19 StreamMoneyBlocked} |
+  --{20 StreamMoneyError} |
+
+  -- DataStream-related frames (hex 0x2?)
+  {32 StreamData} |
+  {33 StreamDataEnd} |
+  --{34 StreamDataMax} |
+  --{35 StreamDataBlocked} |
+  --{36 StreamDataError}
+}
+
+StreamFrame ::= SEQUENCE {
+  type FRAME.&typeId ({FrameSet}),
+  data FRAME.&Type ({PacketSet}{@type})
+}
+
+ConnectionError ::= SEQUENCE {
+  -- Numeric error code
+  code UInt8,
+
+  -- Human-readable error message (UTF-8)
+  message VarBytes
+}
+
+ConnectionNewAddress ::= SEQUENCE {
+  -- ILP Address of the party that sends the frame
+  address Address
+}
+
+StreamMoney ::= SEQUENCE {
+  -- Identifier of the MoneyStream
+  streamId VarUInt,
+
+  -- The proportion of the Prepare amount that should go to this stream
```

Comment:
maybe it's fine, but it seems slightly strange to have this use the sum of all streamMoney frames. You need to keep a running total on the side and then execute the streamMoney logic after all other frames are parsed, rather than being able to apply each frame's effect as you go through them

If we used something like percents or put the sum of all shares into each of the streammoney frames, then you could apply each frame without having to execute money frames at the end.

Reply:
There's no real need to apply it atomically; if the sender makes an error and gives out more than 100% of shares then the last frames just wouldn't be applied to their streams.

Reply:
What if it's < 100% where does the remaining money go?
How does the sender get notified which frames were applied and which weren't?